ffmpeg/lpms_ffmpeg.c

```diff
@@ -944,42 +1233,100 @@ int lpms_transcode(input_params *inp, output_params *params,
-      }
-    }
-    ret = 0;
-    if (octx->ac) { // flush audio
-      while (!ret || ret == AVERROR(EAGAIN)) {
-        ret = process_out(&ictx, octx, octx->ac, octx->oc->streams[octx->ai], &octx->af, NULL);
-      }
-    }
-    av_interleaved_write_frame(octx->oc, NULL); // flush muxer
-    ret = av_write_trailer(octx->oc);
-    if (ret < 0) main_err("transcoder: Unable to write trailer");
+    ret = flush_outputs(ictx, &outputs[i]);
+    if (ret < 0) main_err("transcoder: Unable to fully flush outputs")
+    outputs[i].nb_frame = 0;
   }
 
 transcode_cleanup:
-  free_input(&ictx);
-  for (i = 0; i < MAX_OUTPUT_SIZE; i++) free_output(&outputs[i]);
+  avio_closep(&ictx->ic->pb);
   if (dframe) av_frame_free(&dframe);
+  ictx->flushed = 0;
+  if (ictx->first) av_packet_free(&ictx->first);
+  if (ictx->ac) avcodec_free_context(&ictx->ac);
+  if (ictx->vc && AV_HWDEVICE_TYPE_NONE == ictx->hw_type) avcodec_free_context(&ictx->vc);
+  for (i = 0; i < nb_outputs; i++) free_output(&outputs[i]);
   return ret == AVERROR_EOF ? 0 : ret;
 #undef main_err
 }
+
+int lpms_transcode(input_params *inp, output_params *params,
+  output_results *results, int nb_outputs, output_results *decoded_results)
+{
+  int ret = 0;
+  struct transcode_thread *h = inp->handle;
+
+  pthread_mutex_lock(&h->mu);
+  if (!h->initialized) {
+    int i = 0;
+    int decode_a = 0, decode_v = 0;
+    if (nb_outputs > MAX_OUTPUT_SIZE) {
+      pthread_mutex_unlock(&h->mu);
+      return lpms_ERR_OUTPUTS;
+    }
+
+    // Check to see if we can skip decoding
+    for (i = 0; i < nb_outputs; i++) {
+      if (!needs_decoder(params[i].video.name)) h->ictx.dv = ++decode_v == nb_outputs;
+      if (!needs_decoder(params[i].audio.name)) h->ictx.da = ++decode_a == nb_outputs;
+    }
+
+    h->nb_outputs = nb_outputs;
+
+    // populate input context
+    ret = open_input(inp, &h->ictx);
+    if (ret < 0) {
+      pthread_mutex_unlock(&h->mu);
+      return ret;
+    }
+  }
+
+  if (h->nb_outputs != nb_outputs) {
+    return lpms_ERR_OUTPUTS; // Not the most accurate error...
+    pthread_mutex_unlock(&h->mu);
+  }
+
+  ret = transcode(h, inp, params, results, decoded_results);
+  h->initialized = 1;
+
+  pthread_mutex_unlock(&h->mu);
+
+  return ret;
+}
+
+struct transcode_thread* lpms_transcode_new() {
+  struct transcode_thread *h = malloc(sizeof (struct transcode_thread));
+  if (!h) return NULL;
+  memset(h, 0, sizeof *h);
+  pthread_mutex_init(&h->mu, NULL);
+  return h;
+}
+
+void lpms_transcode_stop(struct transcode_thread *handle) {
+  // not threadsafe as-is; calling function must ensure exclusivity!
+
+  int i;
+
+  if (!handle) return;
+
+  pthread_mutex_lock(&handle->mu);
+  free_input(&handle->ictx);
+  for (i = 0; i < MAX_OUTPUT_SIZE; i++) {
+    free_output(&handle->outputs[i]);
+    if (handle->outputs[i].vc) avcodec_free_context(&handle->outputs[i].vc);
+  }
+  pthread_mutex_unlock(&handle->mu);
+
+  free(handle);
```

Comment:
If `lpms_transcode_stop` called from two threads, isn't this will lead to `free_input(&handle->ictx);` being called with pointer to uninitialized memory?

Reply:
Indeed. Combined with https://github.com/livepeer/lpms/pull/141#discussion_r348263123 and https://github.com/livepeer/lpms/pull/141#discussion_r348028335 , it seems much better to just handle the locking within golang itself. Having a GC helps. Fixed in https://github.com/livepeer/lpms/pull/141/commits/39fe9c5cac114fa0f3ddacdc5c22270f40b65db8